src/common/src/system_param.rs

```diff
@@ -137,10 +131,99 @@ macro_rules! impl_system_params_from_kv {
     };
 }
 
+macro_rules! impl_set_system_param {
+    ($({ $field:ident, $default:expr },)*) => {
+        pub fn set_system_param(params: &mut SystemParams, key: &str, value: Option<String>) -> Result<()> {
+             match key {
+                $(
+                    key_of!($field) => {
+                        let v = if let Some(v) = value {
+                            v.parse().map_err(|_| format!("cannot parse parameter value"))?
+                        } else {
+                            $default
+                        };
+                        sanity_check::$field(&v)?;
+                        params.$field = Some(v);
+                    },
+                )*
+                _ => {
+                    return Err(format!(
+                        "unrecognized system param {:?}",
+                        key
+                    ));
+                }
+            };
+            Ok(())
+        }
+    };
+}
+
 for_all_undeprecated_params!(impl_system_params_from_kv);
 
 for_all_undeprecated_params!(impl_system_params_to_kv);
 
+for_all_undeprecated_params!(impl_set_system_param);
+
+/// Check the validity of set values.
+mod sanity_check {
+    use std::fmt::Debug;
+    use std::ops::{Bound, RangeBounds};
+
+    use super::*;
+
+    pub fn barrier_interval_ms(v: &u32) -> Result<()> {
+        expect_range(*v, (Bound::Excluded(0), Bound::Unbounded))
+    }
+
+    pub fn checkpoint_frequency(v: &u64) -> Result<()> {
```

Comment:
Seems we don't have a concrete type definition for each field and just let the compiler check whether this matches the field type in prost message indirectly?

Reply:
> just let the compiler check whether this matches the field type in prost message indirectly?

I can't find a way to avoid it 🥵